checker: returning value when reference is expected should be reported by V (not C)

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relvant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** 
```
V 0.1.30 d782db8
```
**OS:** 
```
macOS Catalina
```

**What did you do?**
Tried to build my project.


**What did you expect to see?**
To compile my project

 
**What did you see instead?**
```
builder error: 
==================
C error. This should never happen.
```

Full error and source code here: https://gist.github.com/DilsMatchanov/e6ded687394c3bf698d72d53e729e7f2

Now I get this error:
```
==================
                                                       ^  ~~
/var/folders/fq/r919kqjx0995dzmh4twcjg280000gn/T/v/websocketapi.5332331601702602512.tmp.c:29490:89: error: use of undeclared identifier '_t1181'; did you mean '_t1180'?
                int header_len = 2 + /*experimental if expr*/ _t1180 /*Z*/ + /*experimental if expr*/ _t1181 /*Z*/;
                                                                                                      ^~~~~~
                                                                                                      _t1180
/var/folders/fq/r919kqjx0995dzmh4twcjg280000gn/T/v/websocketapi.5332331601702602512.tmp.c:29470:5: note: '_t1180' declared here
int _t1180; /* if prepend */
    ^
/var/folders/fq/r919kqjx0995dzmh4twcjg280000gn/T/v/websocketapi.5332331601702602512.tmp.c:29854:52: error: use of undeclared identifier '_arg_expr_starts_with_0_12040'; did you mean '_arg_expr_starts_with_0_11989'?
                } else if (string_starts_with(/*af receiver arg*/_arg_expr_starts_with_0_12040, _SLIT("wss://"))) {
                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                                 _arg_expr_starts_with_0_11989
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, please raise an issue on GitHub:

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

make: *** [all] Error 1
```

--- a/vlib/live/common.v
+++ b/vlib/live/common.v
@@ -52,7 +52,7 @@ pub mut:
 // so that the user can set callbacks, read meta information, etc.
 pub fn info() &LiveReloadInfo {
 	if C.g_live_info != 0 {
-		return C.g_live_info
+		return &LiveReloadInfo(C.g_live_info)
 	}
 	// When the current program is not compiled with -live, simply
 	// return a new empty struct LiveReloadInfo in order to prevent

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2031,6 +2031,12 @@ pub fn (mut c Checker) return_stmt(mut return_stmt ast.Return) {
 			c.error('fn `$c.cur_fn.name` expects you to return a non reference type `${c.table.type_to_str(exp_type)}`, but you are returning `${c.table.type_to_str(got_typ)}` instead',
 				pos)
 		}
+		if (exp_type.is_ptr() || exp_type.is_pointer()) &&
+			(!got_typ.is_ptr() && !got_typ.is_pointer()) && got_typ != table.any_int_type {
+			pos := return_stmt.exprs[i].position()
+			c.error('fn `$c.cur_fn.name` expects you to return a reference type `${c.table.type_to_str(exp_type)}`, but you are returning `${c.table.type_to_str(got_typ)}` instead',
+				pos)
+		}
 	}
 	if exp_is_optional && return_stmt.exprs.len > 0 {
 		expr0 := return_stmt.exprs[0]

--- a/vlib/v/checker/tests/reference_return.out
+++ b/vlib/v/checker/tests/reference_return.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/reference_return.vv:6:9: error: fn `f` expects you to return a reference type `&Aa`, but you are returning `Aa` instead
+    4 | 
+    5 | fn f() &Aa {
+    6 |     return Aa{
+      |            ~~~
+    7 |         a: 1
+    8 |     }

--- a/vlib/v/checker/tests/reference_return.vv
+++ b/vlib/v/checker/tests/reference_return.vv
@@ -1,0 +1,14 @@
+struct Aa {
+	a int
+}
+
+fn f() &Aa {
+	return Aa{
+		a: 1
+	}
+}
+
+fn main() {
+	x := f()
+	println(x.a)
+}

--- a/vlib/v/tests/ref_return_test.v
+++ b/vlib/v/tests/ref_return_test.v
@@ -1,0 +1,14 @@
+struct Aa {
+	a int
+}
+
+fn f() &Aa {
+	return &Aa{
+		a: 5
+	}
+}
+
+fn test_ref_return() {
+	x := f()
+	assert x.a == 5
+}

